File: Binary Tree - DFS/437. Path Sum III/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number {
    let count = 0;

    const dfs = (node: TreeNode | null, currSum: number, targetSum: number, path: number[]) => {
        if (!node) return;
        currSum += node.val;
        if (currSum === targetSum) count++;
        dfs(node.left, currSum, targetSum, [...path, node.val]);
        dfs(node.right, currSum, targetSum, [...path, node.val]);

        for (let i = 0; i < path.length; i++) {
            currSum -= path[i];
            if (currSum === targetSum) count++;
        }
    };

    dfs(root, 0, targetSum, []);
    return count;
}

File: Binary Tree - DFS/437. Path Sum III/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.count = 0
        
        def dfs(node, curr_sum, target_sum, path):
            if not node:
                return
            curr_sum += node.val
            if curr_sum == target_sum:
                self.count += 1
            dfs(node.left, curr_sum, target_sum, path + [node.val])
            dfs(node.right, curr_sum, target_sum, path + [node.val])
            
            for i in range(len(path)):
                curr_sum -= path[i]
                if curr_sum == target_sum:
                    self.count += 1
        
        dfs(root, 0, targetSum, [])
        return self.count